pkg/store/forward/forward.go

```diff
@@ -0,0 +1,130 @@
+package forward
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/openshift/telemeter/pkg/store"
+	clientmodel "github.com/prometheus/client_model/go"
+	"github.com/prometheus/prometheus/prompb"
+)
+
+const metricName = "__name__"
+
+type Store struct {
+	next   store.Store
+	url    *url.URL
+	client *http.Client
+}
+
+func New(url *url.URL, next store.Store) *Store {
+	return &Store{
+		next:   next,
+		url:    url,
+		client: &http.Client{},
+	}
+}
+
+func (s *Store) ReadMetrics(ctx context.Context, minTimestampMs int64) ([]*store.PartitionedMetrics, error) {
+	return s.next.ReadMetrics(ctx, minTimestampMs)
+}
+
+func (s *Store) WriteMetrics(ctx context.Context, p *store.PartitionedMetrics) error {
+	if p == nil {
+		return nil
+	}
+
+	// Run in a func to catch all transient errors
+	err := func() error {
+		timeseries := convertToTimeseries(p)
+
+		wreq := &prompb.WriteRequest{
+			Timeseries: timeseries,
+		}
+
+		data, err := proto.Marshal(wreq)
+		if err != nil {
+			return err
+		}
+
+		compressed := snappy.Encode(nil, data)
+
+		req, err := http.NewRequest(http.MethodPost, s.url.String(), bytes.NewBuffer(compressed))
+		if err != nil {
+			return err
+		}
+
+		ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+		defer cancel()
+
+		req = req.WithContext(ctx)
+
+		resp, err := s.client.Do(req)
+		if err != nil {
+			return err
+		}
+
+		if resp.StatusCode/100 != 2 {
+			return fmt.Errorf("response was not 200 OK, but %s", resp.Status)
+		}
+
+		return nil
+	}()
+	if err != nil {
+		log.Printf("forwarding error: %v", err)
+	}
+
+	return nil
```

Comment:
If this func will _always_ return a nil error, why not make the entire forwarding func async, that way we can write to memory sooner?

Reply:
Thought about making it async. But not totally sure about the implications of that. What do you think about keeping it like that and then move it to async later, if it becomes a problem?! :thinking:

Reply:
right now the implication of making it synchronous is that writing to memory will be delayed until writing to thanos has succeeded